File: Python3/201 - 400/303 Range Sum Query - Immutable.py
# 60ms, 100.0%
class NumArray:

    def __init__(self, nums):
        if nums:
            self.sums = [nums[0]]
            for i in range(1, len(nums)):
                self.sums.append(self.sums[-1] + nums[i])
        else:
            self.sums = []

    def sumRange(self, i, j):
        if not self.sums: return 0
        if i == 0: return self.sums[j]
        return self.sums[j] - self.sums[i - 1]

File: 201 - 400/303 Range Sum Query - Immutable.cc
// 169ms, 64.54%  ----  WHY runs so slow compared to Python ?
class NumArray {
private:
    int *sums = nullptr;    // will cost 170ms if use vector
    
public:
    NumArray(vector<int> nums) {
        int end = nums.size();
        sums = new int [end + 1];
        sums[0] = 0;
        for (int i = 0; i < end; ++i)
            sums[i + 1] = sums[i] + nums[i];
    }
    
    int sumRange(int i, int j) {
        return sums[j + 1] - sums[i];
    }
};
